_posts/2023-11-10-automated-test-generation-chess-puzzles-with-dafny.markdown

```diff
@@ -0,0 +1,288 @@
+---
+layout: post
+title:  "Automated Test Generation: Chess Puzzles with Dafny"
+date:   2023-11-10 11:00:00 -0500
+author: Sasha Fedchin
+---
+Given all the incredible examples that showcase how powerful Dafny is (did you know that part of Dafny is written in Dafny?), you might think that verification can fully replace testing. And in many cases it does! However if, like a number of industrial Dafny users, you want to integrate Dafny code with existing codebases, you will face challenges that verification alone might not solve:
+
+- Perhaps your code is actually a reference implementation for another program and you want to test the equivalence between the two.
+- You might be calling out to external libraries, which are not themselves verified, and would like to test them by calling your Dafny code.
+- You have written high-level specifications for your code and are about to embark on a proof but want to have some initial assurance that your yet-to-be-verified implementation does conform to these specifications. 
+
+In this blog post, I show how Dafny’s built-in automated test generation can be useful in these situations. In the example I use, Dafny has to position pieces on a chess board in such a way that certain constraints are satisfied. This is exactly what I want to illustrate: chess boards will be generated tests, and constraints will be the program we want to test. A brute force enumeration or fuzzing might not be up to the task here, since the number of ways in which you can arrange pieces on the board is astronomically high, even under the constraints I add to the problem. Dafny, however, can use the verifier to generate tests and is, therefore, much more efficient in finding solutions. 
+
+The post is structured as follow: I will first outline a [subset of chess rules using Dafny](#1-modelling-chess-in-dafny). This code serves as a reference point throughout the post. I then demonstrate [the basics of test generation](#2-test-generation-basics) and discuss the different coverage metrics you can target. This will let me deal with [visualizing coverage and identifying dead code](#3-there-is-dead-code-here). Moving on to advanced features, I will offer a [broader discussion of quantifiers, loops and recursion](#sec-quantifiers-loops-and-recursion) — features that require special care when attempting to generate tests. Finally, this blogpost concludes with a [summary and general guidelines](#conclusions-and-best-practices) for how to apply this technique to your own Dafny programs. You can also find more information on automated test generation in the [Dafny reference manual](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-dafny-generate-tests). If you want to try test generation on your own, I recommend using the latest stable Dafny nightly release.
```

Comment:
> I recommend using the latest stable Dafny nightly release

Can we make the text more future proof by providing some minimal version that's recommended, even if it not yet released? "I recommend using Dafny 4.4 or greater once it's released, and before that the Dafny nightly release"

Reply:
This makes sense! Have added a recommendation to use Dafny 4.4 or above